pkg/controller/certificates/internal/secretsmanager/secret.go

```diff
@@ -59,64 +68,68 @@ type SecretData struct {
-// data as well as appropriate metadata.
-// If the Secret resource does not exist, it will be created.
-// Otherwise, the existing resource will be updated.
-// The first return argument will be true if the resource was updated/created
-// without error.
+// UpdateData will ensure the Secret resource contains the given secret data as
+// well as appropriate metadata using an Apply call.
+// If the Secret resource does not exist, it will be created on Apply.
 // UpdateData will also update deprecated annotations if they exist.
 func (s *SecretsManager) UpdateData(ctx context.Context, crt *cmapi.Certificate, data SecretData) error {
-	// Fetch a copy of the existing Secret resource
-	secret, err := s.secretLister.Secrets(crt.Namespace).Get(crt.Spec.SecretName)
-	if !apierrors.IsNotFound(err) && err != nil {
-		// If secret doesn't exist yet, then don't error
+	secret, err := s.getCertificateSecret(ctx, crt)
+	if err != nil {
 		return err
 	}
-	secretExists := (secret != nil)
 
-	// If the secret does not exist yet, then we need to create one
-	if !secretExists {
-		secret = &corev1.Secret{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      crt.Spec.SecretName,
-				Namespace: crt.Namespace,
-			},
-			Type: corev1.SecretTypeTLS,
-		}
+	log := logf.FromContext(ctx)
+	log = logf.WithResource(log, secret)
+
+	if err := s.setValues(crt, secret, data); err != nil {
+		return err
 	}
 
-	// secret will be overwritten by 'existingSecret' if existingSecret is non-nil
+	// Build Secret apply configuration and options.
+	applyOpts := metav1.ApplyOptions{FieldManager: s.userAgent}
+	applyCnf := applycorev1.Secret(secret.Name, secret.Namespace).
+		WithAnnotations(secret.Annotations).WithLabels(secret.Labels).
+		WithData(secret.Data).WithType(secret.Type)
+
+	// If Secret owner reference is enabled, set it on the Secret. This results
+	// in a no-op if the Secret already exists and has the owner reference set,
+	// and visa-versa.
 	if s.enableSecretOwnerReferences {
-		secret.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(crt, certificateGvk)}
+		ref := *metav1.NewControllerRef(crt, certificateGvk)
+		applyCnf = applyCnf.WithOwnerReferences(&applymetav1.OwnerReferenceApplyConfiguration{
+			APIVersion: &ref.APIVersion, Kind: &ref.Kind,
+			Name: &ref.Name, UID: &ref.UID,
+			Controller: ref.Controller, BlockOwnerDeletion: ref.BlockOwnerDeletion,
+		})
 	}
 
-	secret = secret.DeepCopy()
-	err = s.setValues(crt, secret, data)
-	if err != nil {
-		return err
-	}
+	log.V(logf.DebugLevel).Info("applying secret")
 
-	// If secret does not exist then create it
-	if !secretExists {
+	// Apply secret resource.
+	_, err = s.secretClient.Secrets(secret.Namespace).Apply(ctx, applyCnf, applyOpts)
+	if apierrors.IsConflict(err) {
+		log.Error(err, "forcing apply due to field management conflict")
+		applyOpts.Force = true
+		_, err = s.secretClient.Secrets(secret.Namespace).Apply(ctx, applyCnf, applyOpts)
+	}
```

Comment:
Use the `Force` option always? Or add a comment explaining why that is not desirable.

Reply:
Mainly to have the log output for it. Will add a comment.